pintracker/statelesstracker/statelesstracker.go

```diff
@@ -0,0 +1,510 @@
+
+	select {
+	case spt.pinCh <- c:
+	default:
+		err := errors.New("pin queue is full")
+		spt.setError(c.Cid, err)
+		logger.Error(err.Error())
+		return err
+	}
+	return nil
+}
+
+// Untrack tells the StatelessPinTracker to stop managing a Cid.
+// If the Cid is pinned locally, it will be unpinned.
+func (spt *Tracker) Untrack(c *cid.Cid) error {
+	logger.Debugf("untracking %s", c)
+	if opc, ok := spt.optracker.Get(c); ok {
+		if opc.Op == optracker.OperationPin {
+			spt.optracker.Finish(c) // cancel it
+
+			switch opc.Phase {
+			case optracker.PhaseQueued:
+				return nil
+			case optracker.PhaseInProgress:
+				// continues below to run a full unpin
+			}
+		}
+	}
+
+	spt.optracker.TrackNewOperation(spt.ctx, c, optracker.OperationUnpin)
+
+	select {
+	case spt.unpinCh <- api.PinCid(c):
+	default:
+		err := errors.New("unpin queue is full")
+		spt.setError(c, err)
+		logger.Error(err.Error())
+		return err
+	}
+	return nil
+}
+
+// StatusAll returns information for all Cids pinned to the local IPFS node.
+func (spt *Tracker) StatusAll() []api.PinInfo {
+	// get statuses from ipfs node first
+	localpis, _ := spt.ipfsStatusAll()
+	// get all inflight operations from optracker
+	infops := spt.optracker.GetAll()
+	// put them into the map, deduplicating any already 'pinned' items with
+	// their inflight operation
+	for _, infop := range infops {
+		localpis[infop.Cid.String()] = infop.ToPinInfo(spt.peerID)
+	}
+	// get all remote pins
+	for _, remote := range spt.getRemotesAll() {
+		localpis[remote.Cid.String()] = remote
+	}
+	// get all err pins
+	for _, err := range spt.getErrorsAll() {
+		localpis[err.Cid.String()] = err
+	}
+	// convert to array
+	var pis []api.PinInfo
+	for _, pi := range localpis {
+		logger.Debugf("statusall: pins: %v", pi)
+		pis = append(pis, pi)
+	}
+	return pis
+}
+
+// Status returns information for a Cid pinned to the local IPFS node.
+func (spt *Tracker) Status(c *cid.Cid) api.PinInfo {
+	// check errors map
+	if pi, ok := spt.getError(c); ok {
+		return pi
+	}
+	// check remotes map
+	if pi, ok := spt.getRemote(c); ok {
+		return pi
+	}
+	// check if c has an inflight operation in optracker
+	if op, ok := spt.optracker.Get(c); ok {
+		// if it does return the status of the operation
+		return op.ToPinInfo(spt.peerID)
+	}
+	// else attempt to get status from ipfs node
+	//TODO(ajl): log error
+	pi, _ := spt.ipfsStatus(c)
+	return pi
+}
+
+// SyncAll verifies that the statuses of all tracked Cids (from the shared state)
+// match the one reported by the IPFS daemon. If not, they will be transitioned
+// to PinError or UnpinError.
+//
+// SyncAll returns the list of local status for all tracked Cids which
+// were updated or have errors. Cids in error states can be recovered
+// with Recover().
+// An error is returned if we are unable to contact the IPFS daemon.
+func (spt *Tracker) SyncAll() ([]api.PinInfo, error) {
+	pis := spt.getErrorsAll()
```

Comment:
You should actually make sure that 

1) There are no ongoing Pin operations for things that are already `pinned` in ipfs
2) No ongoing Unpin operations for things already `unpinned` in ipfs
3) No errors for things which have no error (they became pinned or unpinned)